SW-649 create update script for mrbeam plugin

JFSolution opened the pull request:
change the update mechanism of the mrbeam plugin to use a update script and install the plugin dependencies with this script

Commit:
added update script and dependencies

Commit:
fix dependency names

Commit:
save dependencies and update script of target mrbeam version and use these to continue
collect dependencies from dependencies.txt
collect update info from update_info.json
build wheels for venvs
install wheels for venvs

Commit:
errorhandling, refactoring, docstrings

Commit:
add a test for the dependency file

Commit:
refactor get_file_of_repo_for_tag

Commit:
fix wrongly committed change

Commit:
use mrbeam plugin archive zip instead of single files

Commit:
remove comment

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+
+
+def _parse_arguments():
+    import argparse
```
Please move all Imports to the top of the module. it is a recommended best practice. Please see
https://peps.python.org/pep-0008/#imports
https://stackoverflow.com/a/1025259

JFSolution replied:
ok, will do

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+def _parse_arguments():
+    import argparse
+
+    boolean_trues = ["true", "yes", "1"]
```
why we need this ?

JFSolution replied:
I copied this from the octoprint update script, this will accept `true` `yes` and `1` as input

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+
+    boolean_trues = ["true", "yes", "1"]
+
+    parser = argparse.ArgumentParser(prog="update_script.py")
```
could this "update_script.py" be replace with `__file__` ?

JFSolution replied:
yes

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+        default=None,
+        help="Specify the branch to make sure is checked out",
+    )
+    parser.add_argument(
```
help is missing here

JFSolution replied:
I added

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+        type=str,
+        dest="archive",
+        default=None,
+        help="path of target zip file on local system",
```
typo: `path` --> `Path`

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+        type=str,
+        dest="target_version",
+        default=None,
+        help="version number of the target",
```
typo: `version` --> `Version`

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+        help="version number of the target",
+    )
+    parser.add_argument(
+        "folder",
```
missing `--` or is it intended ?

JFSolution replied:
this is intended so it is a required param

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+        type=str,
+        help="Specify the base folder of the OctoPrint installation to update",
+    )
+    parser.add_argument(
```
missing `--` or is it intended ?

JFSolution replied:
this is intended so it is a required param

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+    except IOError:
+        raise RuntimeError("Could not load update info")
+    except ValueError:
+        raise RuntimeError("update info not valid json")
```
you mean valid yaml ?

JFSolution replied:
no it is a json file

ahmed-mrbeam replied:
Then that is confusing as you use yaml to load it's content

JFSolution replied:
I will try with json.load()

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+    except ValueError:
+        raise RuntimeError("update info not valid json")
+    except yaml.YAMLError as e:
+        raise RuntimeError("update info not valid json - {}".format(e))
```
you mean yaml, right ? also why not combining this except with the one before ?

JFSolution replied:
it is a json file thats why I would show this message, I moved them together

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+    return update_info
+
+
+def build_wheels(queue):
```
please change the parameter name as it is pretty similar to the queue class in python

JFSolution replied:
ok

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+    """
+    for venv, packages in queue.items():
+
+        pip_caller = _get_pip_caller(command=venv)
```
can we use this one instead ?
`can_perform_update` ?
can be found in:
`octoprint.plugins.softwareupdate.updaters.pip`

JFSolution replied:
can_performe_update will only return true or false, I need the pip caller itself

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+        if pip_caller is None:
+            raise exceptions.UpdateError("Can't run pip", None)
+
+        def _log_call(*lines):
```
do we actually need to define these methods twice ?, one here and another time in install wheels?
I recommend to move them out , even to the utils file if suitable as they seem Generic

JFSolution replied:
good point

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+                pip_args.append(package.get("archive"))
+            else:
+                raise exceptions.UpdateError(
+                    "Archive not found for package {}".format(package)
```
This will dump the whole dict, right ?, do we need that or do you just intended to print the name?

JFSolution replied:
nothing will be dumped here, I only append the name of the archive to the pip args, so that the wheels will be build for the archive

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+                    "Archive not found for package {}".format(package)
+                )
+
+        returncode, stdout, stderr = pip_caller.execute(*pip_args)
```
a little bit not a fan of using variable names that are similar to keywords like `stdout ` and `stderr`, what do you think ?

JFSolution replied:
I prepende exec_

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+            )
+
+
+def install_wheels(queue):
```
same comment on using param name `queue`

JFSolution replied:
ok changed to install_queue

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+    Returns:
+        None
+    """
+    for venv, packages in queue.items():
```
what if queue is not a dictionary ?

JFSolution replied:
I add a exception for this

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+    for venv, packages in queue.items():
+
+        pip_caller = _get_pip_caller(command=venv)
+        if pip_caller is None:
```
better to use:
`if  isinstance(pip_caller, PipCaller) is False`

JFSolution replied:
ok

Commit:
refactor update script to easily reuse it in netconnectd

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -0,0 +1,505 @@
+        }
+    )
+
+    if dependencies:
```
Please add some info logs so that we would know if there are dependencies, and what are they. and to know also if there are no dependencies

JFSolution replied:
ok

JFSolution replied:
the output will show like this now
```
dependencies - [{'version': '0.7.15', 'name': 'iobeam'}, {'version': '0.0.25', 'name': 'mrb-hw-info'}, {'version': '0.2.2', 'name': 'mrbeam-ledstrips'}, {'version': '1.0.0-alpha', 'name': 'mrbeamdoc'}]
skip dependency iobeam as the target version 0.7.15 is already installed
skip dependency mrb-hw-info as the target version 0.0.25 is already installed
skip dependency mrbeam-ledstrips as the target version 0.2.2 is already installed
install_queue {'/home/pi/oprint/bin/pip': [{'target': '0.10.3', 'name': 'Mr_Beam', 'archive': '/tmp/octoprint/mrbeamplugin/2b1a16967aac314a888a9c090957def5902b2679.zip'}, {'target': '1.0.0-alpha', 'name': 'mrbeamdoc', 'archive': u'https://github.com/mrbeam/MrBeamDoc/archive/v1.0.0-alpha.zip'}]}
```

Commit:
change dependency regex to symantic versioning
change software update generation logic for update scripts
add new error notification for errors during fetch update information

ahmed-mrbeam commented on octoprint_mrbeam/scripts/update_script.py:
```diff
@@ -125,7 +125,7 @@ def get_dependencies(path):
     """
     dependencies_path = os.path.join(path, "dependencies.txt")
-    dependencies_pattern = r"([a-z]+(?:[_-][a-z]+)*)(.=)+([0-9]+.[0-9]+.[0-9]+)"
+    dependencies_pattern = r"([a-z]+(?:[_-][a-z]+)*)(.=)+((0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?)"
```
Please put a comment with example of the input string and the output after the match

JFSolution replied:
ok

ahmed-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -366,13 +366,36 @@ def _generate_config_of_module(
-                plugin._basefolder, "scripts", "update_script.py"
+        if "update_script" in input_moduleconfig:
+            if "update_script_relative_path" not in input_moduleconfig:
+                _logger.error(
```
so you only need to log the error and continue normally ?

JFSolution replied:
hmm good question, what do you think we should do here? We can show a user notification and return none for the update_config

ahmed-mrbeam replied:
User notification for update failed, right ?

JFSolution replied:
I added a notification `Error during fetching update information` with errorcode E-1003 and will break the fetching methode

ahmed-mrbeam commented on tests/softwareupdate/test_dependencies.py:
```diff
@@ -0,0 +1,16 @@
+import unittest
+
+
+class UpdateScriptTestCase(unittest.TestCase):
```
`UpdateScriptTestCase` --> `TestUpdateScript`

ahmed-mrbeam replied:
why the name test case ?

JFSolution replied:
because it ingertis from TestCase, but I will change it to you propose

ahmed-mrbeam commented on octoprint_mrbeam/util/github_api.py:
```diff
@@ -0,0 +1,56 @@
+# TODO move to util an refactor cloud config part to use same
```
is the comment still needed ?

JFSolution replied:
no, you are right

ahmed-mrbeam commented on octoprint_mrbeam/util/github_api.py:
```diff
@@ -0,0 +1,56 @@
+# TODO move to util an refactor cloud config part to use same
+import base64
```
Please add a proper doc string at the the top of the file

JFSolution replied:
What is a proper doc string for such files? We didn't use them so far

ahmed-mrbeam replied:
I believe when I started reviewing there was no doc string here at the top. Now this one is OK

ahmed-mrbeam commented on octoprint_mrbeam/software_update_information.py:
```diff
@@ -176,7 +132,13 @@ def get_update_information(plugin):
+                    get_file_of_repo_for_tag(
+                        repo="beamos_config",
+                        file="docs/sw-update-conf.json",
+                        tag="v{tag}".format(tag=str(config_tag)),
```
don't we already name the tag with v?

JFSolution replied:
yes, but it will be removed from the semantic versioning select. But it makes sense to prepend it afterwards

Commit:
added requested changes
 -refactoring
 -move get_pip_caller to pip_util
 -additional logging

Commit:
added a example for the regex
fixed folder creation

ahmed-mrbeam commented on octoprint_mrbeam/util/pip_util.py:
```diff
@@ -62,3 +64,31 @@ def get_version_of_pip_module(pip_name, pip_command=None, disable_pip_ver_check=
     return version
+
+
+def get_pip_caller(venv, _logger=None):
```
Doc string missing

JFSolution replied:
will add

Commit:
added doc string https://github.com/mrbeam/MrBeamPlugin/pull/1438/files#r832101702

Commit:
added user notification https://github.com/mrbeam/MrBeamPlugin/pull/1438#discussion_r829789281

Commit:
changes needed to run with versioneer versions
- manually create /tmp/wheelhose folder
- delete temp venv wheel folder before creating the wheels
- remove target version of the plugin

Commit:
adapt update tests for new cloud config

Commit:
remove the default config merge from the dependencies
